src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,244 @@
+                SyntaxKind.WhileStatement,
+                SyntaxKind.TryStatement,
+                SyntaxKind.UsingStatement,
+                SyntaxKind.YieldReturnStatement);
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpInlineMethodRefactoringProvider() : base(CSharpSyntaxFacts.Instance, CSharpExpressionPrecedenceService.Instance)
+        {
+        }
+
+        private static bool ShouldStatementBeInlined(StatementSyntax statementSyntax)
+            => statementSyntax is ReturnStatementSyntax || statementSyntax is ExpressionStatementSyntax;
+
+        protected override bool IsSingleStatementOrExpressionMethod(SyntaxNode calleeMethodDeclarationSyntaxNode)
+        {
+            if (calleeMethodDeclarationSyntaxNode is MethodDeclarationSyntax declarationSyntax)
+            {
+                var blockSyntaxNode = declarationSyntax.Body;
+                // 1. If it is an ordinary method with block
+                if (blockSyntaxNode != null)
+                {
+                    var blockStatements = blockSyntaxNode.Statements;
+                    return blockStatements.Count == 1 && ShouldStatementBeInlined(blockStatements[0]);
+                }
+                else
+                {
+                    // 2. If it is an Arrow Expression
+                    var arrowExpressionNodes = declarationSyntax.ExpressionBody;
+                    return arrowExpressionNodes != null;
+                }
+            }
+
+            return false;
+        }
+
+        protected override IParameterSymbol? GetParameterSymbol(SemanticModel semanticModel,ArgumentSyntax argumentSyntaxNode, CancellationToken cancellationToken)
+            => argumentSyntaxNode.DetermineParameter(semanticModel, allowParams: true, cancellationToken);
+
+        protected override SyntaxNode? GetInlineStatement(SyntaxNode calleeMethodDeclarationSyntaxNode)
+        {
+            var declarationSyntax = (MethodDeclarationSyntax)calleeMethodDeclarationSyntaxNode;
+            SyntaxNode? inlineSyntaxNode = null;
+            var blockSyntaxNode = declarationSyntax.Body;
+            // 1. If it is a ordinary method with block
+            if (blockSyntaxNode != null)
+            {
+                var blockStatements = blockSyntaxNode.Statements;
+                if (blockStatements.Count == 1)
+                {
+                    inlineSyntaxNode = GetExpressionFromStatementSyntaxNode(blockStatements[0]);
+                }
+
+                return inlineSyntaxNode;
+            }
+            else
+            {
+                // 2. If it is using Arrow Expression
+                var arrowExpressionNode = declarationSyntax.ExpressionBody;
+                if (arrowExpressionNode != null)
+                {
+                    inlineSyntaxNode = arrowExpressionNode.Expression;
+                }
+
+                return inlineSyntaxNode;
+            }
```

Comment:
i would extract this into a helper.  I woudl then use the helper from IsSingleStatementOrExpressionMethod to basically say that it is a is one of those methods if you can grab this single expression.  That will keep you fro mneeding to repeat the genereral pattern of breaking down the statemetns and grabbing the exprs.